utils/treesitter/grammar.js

```diff
@@ -0,0 +1,407 @@
+    deref_expression: ($) =>
+      prec.right(PREC.TermPrefix, seq('*', $._expression)),
+    fn_type_expression: ($) =>
+      prec.left(seq('fn', $.tuple, '->', $._expression)),
+
+    if_expression: ($) =>
+      prec(
+        PREC.If,
+        seq('if', $._expression, 'then', $._expression, 'else', $._expression)
+      ),
+
+    paren_expression: ($) =>
+      seq('(', repeat_sep($._expression, ','), optional(','), ')'),
+    tuple: ($) => $.paren_expression,
+
+    index_expression: ($) =>
+      prec(PREC.TermPostfix, seq($._expression, '[', $._expression, ']')),
+    designator: ($) => seq('.', choice('base', $.ident)),
+    postfix_expression: ($) =>
+      prec(
+        PREC.TermPostfix,
+        seq(
+          $._expression,
+          choice(
+            '++',
+            '--',
+            $.designator,
+            seq('->', $.ident),
+            seq(choice('.', '->'), '(', $._expression, ')')
+          )
+        )
+      ),
+    where_clause: ($) =>
+      prec(
+        PREC.Where,
+        choice(
+          seq($._expression, '==', $._expression),
+          seq($._expression, 'impls', $._expression),
+          seq($.designator, '=', $._expression),
+          prec.left(seq($.where_clause, choice('or', 'and'), $.where_clause))
+        )
+      ),
+    where_expression: ($) =>
+      prec.left(PREC.TermPostfix, seq($._expression, 'where', $.where_clause)),
+    call_expression: ($) => prec(PREC.TermPostfix, seq($._expression, $.tuple)),
+    pointer_expression: ($) =>
+      prec(PREC.TypePostfix, seq($._expression, $.postfix_star)),
+    _expression: ($) =>
+      choice(
+        $.ident,
+        $.pointer_expression,
+        $.index_expression,
+        $.fn_type_expression,
+        $.designator,
+        seq('.', 'Self'),
+        $.literal,
+        $.builtin_type,
+        $.paren_expression,
+        $.array_literal,
+        $.if_expression,
+        $.binary_expression,
+        $.unary_prefix_expression,
+        $.as_expression,
+        $.paren_expression,
+        $.postfix_expression,
+        $.call_expression,
+        $.ref_expression,
+        $.deref_expression,
+        $.where_expression
+      ),
+
+    _variable_declaration_inner: ($) => seq($.ident, ':', $._expression),
+    variable_declaration: ($) =>
+      choice(
+        seq('var', $.pattern, optional(seq('=', $._expression)), ';'),
+        seq('let', $.pattern, '=', $._expression, ';')
+      ),
+
+    clause: ($) =>
+      seq(choice(seq('case', $.pattern), 'default'), '=>', $.block),
+    clause_list: ($) => seq($.clause_list, $.clause),
+
+    assign_statement: ($) =>
+      seq($._expression, $._assign_operator, $._expression, ';'),
+
+    _assign_operator: ($) =>
+      choice('=', '+=', '/=', '*=', '%=', '-=', '&=', '|=', '^=', '<<=', '>>='),
+
+    match_statement: ($) =>
+      seq('match', '(', $._expression, ')', '{', $.clause_list, '}'),
+    returned_var_statement: ($) =>
```

Comment:
Blank lines are used inconsistently. Either always have blank lines between rules, or only omit them between closely related rules. `assign_statement` is more closely related to `_assign_operator` than `match_statement` is to `returned_var_statement`.